### PR TITLE
Feat/workspace categories settings

### PR DESCRIPTION
**Pull request by ArekChr**
### Details
This PR addresses the implementation of the categories settings page, allowing to enable or disable the expense category requirement.

### Fixed Issues

$ https://github.com/Expensify/App/issues/35706
PROPOSAL: https://docs.google.com/document/d/1gk3xqOs7epMbUrSSiX8K7YcqfPLVgqEos0sf-D-GMDA/edit#bookmark=id.2cvhhb9mrm8w

### Tests

- [x] Verify that no errors appear in the JS console
Pre-requisite: an account in the violations beta

1. Navigate to the workspace settings as an admin within a paid policy group.
2. Verify that you can access the workspace categories settings page.
3. Verify that you can disable and enable "categorize all spend" option.
4. Access workspace settings as a member/admin with a free group policy or member of paid policy group
5. Replace the sub-path /profile with /categories/settings in the URL.
6. Confirm that you are unable to access the categories page.
7. Navigate to `Workspace Settings > Collect Workspace > Categories > Settings`
8. Verify that Members must categorize all spend is off
9. Navigate to the Workspace chat for that Workspace
10. `+ > Request money and verify that Category` is not required
11. Navigate to Workspace Settings > Collect Workspace > Categories > Settings
12. Turn on Members must categorize all spend
13. Repeat step 4 and verify that category shows up as required (only shown if account is in the violations beta). Note that we can still submit the request without adding a category, this is intended as we don't block the user and instead show a violation after the request is made.
14. Make the request, verify that there's a RBR icon next to the request and the category shows an error.

https://github.com/Expensify/App/assets/22219519/846e2889-f06a-4611-b105-308c4351f7ec

### Offline tests
1. Switch network off
2. Go to workspace with paid policy group as a admin
3. Go to workspace categories settings 
4. toggle on or off "members must categorize all spend"
5. Switch network on
6. Verify that toggle has changed and is not "disabled"

### QA Steps
Same as tests

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message: https://expensify.slack.com/archives/C01GTK53T8Q/p1708696592397729
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

Zrzut ekranu 2024-02-23 o 14 30 10
Zrzut ekranu 2024-02-23 o 14 30 18

</details>

<details>
<summary>Android: mWeb Chrome</summary>

Zrzut ekranu 2024-02-23 o 14 33 48
Zrzut ekranu 2024-02-23 o 14 33 41

</details>

<details>
<summary>iOS: Native</summary>

Zrzut ekranu 2024-02-23 o 14 30 14
Zrzut ekranu 2024-02-23 o 14 30 12

</details>

<details>
<summary>iOS: mWeb Safari</summary>

Zrzut ekranu 2024-02-23 o 14 31 25
Zrzut ekranu 2024-02-23 o 14 31 27

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/24796318/458eff66-9ade-400f-a4fa-b3f03c4aa997

Zrzut ekranu 2024-02-23 o 14 38 58

</details>

<details>
<summary>MacOS: Desktop</summary>

Zrzut ekranu 2024-02-23 o 14 11 05

</details>

### CODE REVIEW

**Comment by allroundexperts** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Review by luacmartins — DISMISSED** [review]
Nice, changes LGTM. Did some basic testing and things seem to work well. I'm gonna add the following to the test steps:

1. Navigate to `Workspace Settings > Collect Workspace > Categories > Settings`
2. Verify that `Members must categorize all spend` is off
3. Navigate to the Workspace chat for that Workspace
4. `+ > Request money and verify that `Category` is not required
5. Navigate to `Workspace Settings > Collect Workspace > Categories > Settings`
6. Turn on `Members must categorize all spend`
7. Repeat step 4 and verify that category shows up as required. Note that we can still submit the request without adding a category, this is intended as we don't block the user and instead show a violation after the request is made.
8. Make the request, verify that there's a RBR icon next to the request and the category shows an error.

https://github.com/Expensify/App/assets/22219519/4cb02e11-63cd-4c7a-8f26-3dfd8573282e

**Comment by allroundexperts** [comment]
**BUG**

Setting the toggle optimistically does not result in the category option showing up as required.

Uploading Screen Recording 2024-02-23 at 11.11.38 PM.mov…

**Comment by allroundexperts** [comment]
NitPick: I don't see `You are offline` message being shown anywhere else on the RHN. Are we sure that we want to show it here?

Screenshot 2024-02-23 at 11 10 54 PM

**Comment by luacmartins** [comment]
@ArekChr we also have conflicts now

**Comment by allroundexperts** [comment]
Even when I am online, the `required` flag is not showing up for me. I'm using a paid policy with id `1F27FEAE0B1064E5`.

https://github.com/Expensify/App/assets/30054992/ded4fdde-d723-4c1a-a4c6-b601cbde4082

**Comment by luacmartins** [comment]
@allroundexperts oh that's weird. I definitely saw the required flag (see my video in the OP). Is the call to `SetWorkspaceRequiresCategoryParams` returning an error for you?

**Comment by ArekChr** [comment]
Conflicts resolved

**Comment by ArekChr** [comment]
> NitPick: I don't see `You are offline` message being shown anywhere else on the RHN. Are we sure that we want to show it here?

You're right, last time i saw in year of birth page but it seems it has been removed, fixing

**Comment by cubuspl42** [comment]
Will this require a double C+ review? I see that @allroundexperts started working on this

File: src/pages/workspace/categories/WorkspaceCategoriesSettingsPage.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    return (
+        <AdminPolicyAccessOrNotFoundWrapper policyID={route.params.policyID}>
+            <PaidPolicyAccessOrNotFoundWrapper policyID={route.params.policyID}>
+                {({policy}) => (
+                    <ScreenWrapper
+                        includeSafeAreaPaddingBottom={false}
+                        style={[styles.defaultModalContainer]}
+                        testID={WorkspaceCategoriesSettingsPage.displayName}
+                        shouldShowOfflineIndicatorInWideScreen
```
Comment (luacmartins) [inline src/pages/workspace/categories/WorkspaceCategoriesSettingsPage.tsx:34]: @ArekChr I'm not sure that the last commit was correct. I can still see the offline message. I think we should revert that so we still have the `OfflineWithFeedback` component to apply the offline/optimistic styles to the switch and instead just remove this line
```suggestion
```

**Review by luacmartins — CHANGES_REQUESTED** [review]
Left a comment about the offline styles. Everything else looks good to me. 

I'm not seeing the issue with the `required` flag that @allroundexperts mentioned. It works for me both online and offline.


https://github.com/Expensify/App/assets/22219519/718877a0-335e-423f-97d8-61f169fa03f1

**Comment by luacmartins** [comment]
@cubuspl42 I think we can let @allroundexperts handle this once since they already got started. Thanks for checking though

**Comment by shawnborton** [comment]
Is this one ready for a Design review?

**Comment by luacmartins** [comment]
@shawnborton there's a [known offline](https://github.com/Expensify/App/pull/37133#issuecomment-1961788561) issue. Additionally we're [not applying pattern B styles to the switch](https://github.com/Expensify/App/pull/37133#discussion_r1503177848). Other than that, please go ahead with the review.

**Comment by allroundexperts** [comment]
@ArekChr Are you waiting on translations for the checklist?

**Comment by allroundexperts** [comment]
On a Paid workspace whose trial is expired, we're showing the option to enforce categories but when requesting money from the workspace, the category does not show up as required. @ArekChr Can you maybe check why is this the case?

File: src/pages/workspace/categories/WorkspaceCategoriesSettingsPage.tsx (hunk 1 above)
Comment (ArekChr) [reply src/pages/workspace/categories/WorkspaceCategoriesSettingsPage.tsx:36]: You are right, the name of the component confused me, fixing!

**Comment by ArekChr** [comment]
> @ArekChr Did you manage to find a solution for [#37133 (comment)](https://github.com/Expensify/App/pull/37133#issuecomment-1961787698)?

I have the same issue when setting requires category to true, category is still not required in manual request. Checked that and this is because requires category flag is set to require a logical AND condition together with the `canUseViolations` flag, I'm not pretty sure what is `canUseViolations`. cc @luacmartins

**Comment by ArekChr** [comment]
> @ArekChr Are you waiting on translations for the checklist?

Translations reviewed, all good!

**Comment by ArekChr** [comment]
> On a Paid workspace whose trial is expired, we're showing the option to enforce categories but when requesting money from the workspace, the category does not show up as required. @ArekChr Can you maybe check why is this the case?

The same as above, I think this relates to `canUseViolations` flag from permissions

**Comment by ArekChr** [comment]
> Cool, I agree with the prior feedback that the Offline indicator should be at the very bottom of the view.
> 
> Otherwise this screen looks good to me: CleanShot 2024-02-26 at 21 22 24@2x

I'm confused again, so do we want to show it at very bottom or remove at all?

Double checked it, we don't show offline indicator in other RHP pages

**Comment by luacmartins** [comment]
>I have the same issue when setting requires category to true, category is still not required in manual request. Checked that and this is because requires category flag is set to require a logical AND condition together with the canUseViolations flag, I'm not pretty sure what means canUseViolations

Nice find @ArekChr. That's just a check to verify that the user is part of the violations beta, but we can remove that and always show the `required` flag if `requiresCategory = true`

**Review by luacmartins — APPROVED** [review]
LGTM!

**Review by allroundexperts — APPROVED** [review]
Looks good other than the NAB.

**Comment by luacmartins** [comment]
Hmm actually @JmillsExpensify just mentioned that we should keep that beta check for now. So I'm just gonna update the test steps and I think we're good to merge.

**Review by luacmartins — APPROVED** [review]

**Comment by ArekChr** [comment]
🚀

**Pull request merged**